File: djtoolkit/api/audit.py
"""Fire-and-forget audit logging for sensitive API operations.

Usage::

    from djtoolkit.api.audit import audit_log

    await audit_log(
        user_id=user.user_id,
        action="track.import.csv",
        resource_type="track",
        details={"imported": 42},
        ip_address=request.client.host,
    )

Inserts are performed directly via the pool (service role, bypasses RLS).
Failures are logged as warnings and never block the calling request.
"""

from __future__ import annotations

import json
import logging
from typing import Any

log = logging.getLogger(__name__)


async def audit_log(
    user_id: str,
    action: str,
    *,
    resource_type: str | None = None,
    resource_id: str | None = None,
    details: dict[str, Any] | None = None,
    ip_address: str | None = None,
) -> None:
    """Insert an audit log row.  Fire-and-forget — never raises."""
    try:
        from djtoolkit.db.postgres import get_pool

        pool = await get_pool()
        await pool.execute(
            """
            INSERT INTO audit_logs (user_id, action, resource_type, resource_id, details, ip_address)
            VALUES ($1, $2, $3, $4, $5::jsonb, $6)
            """,
            user_id,
            action,
            resource_type,
            resource_id,
            json.dumps(details) if details else None,
            ip_address,
        )
    except Exception:
        log.warning("audit_log failed for action=%s user=%s", action, user_id, exc_info=True)

File: web/lib/api-server/audit.ts
/**
 * Fire-and-forget audit logging for sensitive API operations.
 *
 * Usage:
 *
 *   import { auditLog, getClientIp } from "@/lib/api-server/audit";
 *
 *   await auditLog(userId, "track.import.csv", {
 *     resourceType: "track",
 *     details: { imported: 42 },
 *     ipAddress: getClientIp(request),
 *   });
 *
 * Inserts via the service-role Supabase client (bypasses RLS).
 * Failures are swallowed and never block the calling request.
 */

import { createServiceClient } from "@/lib/supabase/service";

export interface AuditLogOptions {
  resourceType?: string;
  resourceId?: string;
  details?: Record<string, unknown>;
  ipAddress?: string | null;
}

/**
 * Insert an audit log row. Fire-and-forget — never throws.
 */
export async function auditLog(
  userId: string,
  action: string,
  opts: AuditLogOptions = {}
): Promise<void> {
  try {
    const supabase = createServiceClient();
    const { error } = await supabase.from("audit_logs").insert({
      user_id: userId,
      action,
      resource_type: opts.resourceType ?? null,
      resource_id: opts.resourceId ?? null,
      details: opts.details ?? null,
      ip_address: opts.ipAddress ?? null,
    });
    if (error) {
      console.warn(
        `audit_log failed for action=${action} user=${userId}:`,
        error.message
      );
    }
  } catch (err) {
    console.warn(
      `audit_log threw for action=${action} user=${userId}:`,
      err
    );
  }
}

/**
 * Extract the client IP address from incoming request headers.
 * Respects x-forwarded-for (Vercel/proxy) and x-real-ip, falling back to null.
 */
export function getClientIp(request: Request): string | null {
  const forwarded = request.headers.get("x-forwarded-for");
  if (forwarded) {
    // x-forwarded-for may be a comma-separated list; first entry is the client
    return forwarded.split(",")[0].trim();
  }
  return request.headers.get("x-real-ip");
}
